core/context-strategy/autoscale-context.ts

```diff
@@ -106,87 +132,532 @@ export class PreferredGroupPrimaryElection implements PrimaryElectionStrategy {
+        let electedPrimaryId: string;
+        let highestScore = 0;
+        let multipleHighest = false;
+        scores.forEach((score, vmId) => {
+            if (score > highestScore) {
+                highestScore = score;
+                electedPrimaryId = vmId;
+                multipleHighest = false;
+            } else if (score === highestScore) {
+                multipleHighest = true;
+            }
+        });
+        let electedPrimaryHealthCheckRecord: HealthCheckRecord;
+        if (highestScore > 0 && !multipleHighest) {
+            [electedPrimaryHealthCheckRecord] = allHealthCheckRecords.filter(
+                rec => rec.vmId === electedPrimaryId
+            );
+        }
+
+        // if no primary can be determined using the weighted methods, try the special method.
+        // NOTE: special condition. usually occurs in the intial state where all VM are initially
+        // launched as a secondary role, never sync with any other.
+        // in this case whichever has the oldest send_time will be the election primary
+        // run the method to find the one
+        if (!electedPrimaryHealthCheckRecord) {
+            electedPrimaryId = this.weightMethodSpecial(scores, allHealthCheckRecords);
+            [electedPrimaryHealthCheckRecord] = allHealthCheckRecords.filter(
+                rec => rec.vmId === electedPrimaryId
+            );
+        }
+
+        // still cannot determine the primary, election strategy ends.
+        if (!electedPrimaryHealthCheckRecord) {
+            this.res.newPrimary = null;
+            this.res.newPrimaryRecord = null;
+            return PrimaryElectionStrategyResult.CannotDeterminePrimary;
+        }
+
+        const primaryVm = await this.platform.getVmById(electedPrimaryHealthCheckRecord.vmId);
+
+        // if cannot find the vm, which is rare, do not set this.env.newPrimary
+        this.res.newPrimary = primaryVm || null;
+
+        const primaryRecord: PrimaryRecord = {
+            id: `${electedPrimaryHealthCheckRecord.scalingGroupName}:${electedPrimaryHealthCheckRecord.vmId}`,
+            ip: electedPrimaryHealthCheckRecord.ip,
+            vmId: electedPrimaryHealthCheckRecord.vmId,
+            scalingGroupName: electedPrimaryHealthCheckRecord.scalingGroupName,
+            virtualNetworkId: (primaryVm && primaryVm.virtualNetworkId) || '',
+            subnetId: (primaryVm && primaryVm.subnetId) || '',
+            voteEndTime: Date.now(), // election ends immediately,
+            voteState: PrimaryRecordVoteState.Done
+        };
+
+        this.res.newPrimaryRecord = primaryRecord;
+        return PrimaryElectionStrategyResult.CompleteAndContinue;
+    }
+
+    /**
+     * get (1) point if a record has non-numm checksum
+     * @param {HealthCheckRecord} rec the healthcheck record to earn points
+     * @returns {number} point
+     */
+    weightMethod1(rec: HealthCheckRecord): number {
+        return (rec.deviceChecksum !== null && 1) || 0;
```

Comment:
Prefer `rec.deviceChecksum !== null ? 1 : 0`

misusing boolean operators instead is confusing.

Reply:
fixed